src/core/execute.c

```diff
@@ -4158,8 +4158,29 @@ static int exec_child(
                 return log_oom();
         }
 
+        /* The PATH variable is set to the default path in params->environment.
+         * However, this is overridden if user specified fields have PATH set.
+         * The intention is to also override PATH if the user does
+         * not specify PATH and the user has specified ExecSearchPath
+         */
+
+        if (!strv_isempty(context->exec_search_path)) {
+                _cleanup_free_ char *joined = NULL;
+
+                joined = strv_join(context->exec_search_path, ":");
+                if (!joined)
+                        return log_oom();
+
+                r = strv_env_assign(&joined_exec_search_path, "PATH", TAKE_PTR(joined));
```

Comment:
You leak `joined` here. strv_env_assign() doesn#t take possession of the strings you pass, and you invalidate `joined` via TAKE_PTR() here, hence you leak it because nothing references it anymore then. Just drop the TAKE_PTR()